pinot-controller/src/main/java/com/linkedin/pinot/controller/api/resources/PinotSegmentUploadRestletResource.java

```diff
@@ -413,27 +415,124 @@ private PinotResourceManagerResponse uploadSegment(File indexDir, SegmentMetadat
     }
 
     String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(segmentMetadata.getTableName());
-    OfflineSegmentZKMetadata offlineSegmentZKMetadata = _pinotHelixResourceManager.getOfflineSegmentZkMetadata(
-        offlineTableName, segmentName);
+    ZNRecord znRecord = _pinotHelixResourceManager.getSegmentMetadataZnRecord(offlineTableName, segmentName);
 
-    if (offlineSegmentZKMetadata == null) {
+    if (znRecord == null) {
       // Creating new segment, not refreshing
       downloadUrl = moveSegmentToPermanentDirectory(provider, tableName, segmentName, tempTarredSegmentFile);
       _pinotHelixResourceManager.addNewSegment(segmentMetadata, downloadUrl);
       response = PinotResourceManagerResponse.SUCCESS_RESPONSE;
       return response;
     }
-    if (downloadUrl == null) {
-      downloadUrl = moveSegmentToPermanentDirectory(provider, tableName, segmentName, tempTarredSegmentFile);
+
+    final OfflineSegmentZKMetadata originalOfflineSegmentZkMetadata = new OfflineSegmentZKMetadata(znRecord);
+    OfflineSegmentZKMetadata offlineSegmentZKMetadataWithPushStatus = new OfflineSegmentZKMetadata(znRecord);
+
+    long minionExpectedCrc = -1L;
+    boolean isMinionConfigured = offlineTableConfig.getTaskConfig() != null;
+    boolean minionHeaderPresent = headers.getHeaderString(HttpHeaders.USER_AGENT).contains(CommonConstants.Minion.MINION_HEADER);
+    try {
+      if (isMinionConfigured) {
+        // Minion configured use case so that we write to zk sparingly
+        String segmentPushStatus = originalOfflineSegmentZkMetadata.getSegmentPushStatus();
+        SimpleDateFormat sdf = new SimpleDateFormat("yyyyMMdd'T'HH:mm:ss.SSS");
+        Date dt = new Date();
+        String currentHumanReadableTime = sdf.format(dt);
+        offlineSegmentZKMetadataWithPushStatus.setSegmentPushStatus("COMMITTING-" + currentHumanReadableTime);
+
+        if (segmentPushStatus != null) {
+          if (tooLate(dt)) {
+            LOGGER.error("Segment was not properly committed, replacing segment {} for table {}", segmentName,
+                tableName);
+            _controllerMetrics.addMeteredGlobalValue(ControllerMeter.CONTROLLER_SEGMENT_UPLOAD_ERROR, 1L);
```

Comment:
You should bump a different (new) metric. ControllerMeter.NUMBER_SEGMENT_COMMIT_TIMEOUT_EXCEEDED. We will have alerts on this one.